---
title: "Data Analysis 3.21.17"
author: "Kaiti"
date: "March 21, 2017"
output: pdf_document
---

# Ensuring files are accessible
```{r}
setwd("/home/kaiti/Documents/R Data")
library(GenomicRanges)

# gr.bound is the set of all TADs shared by hESCs and IMR90s as defined by Dixon (2012)
TADbound = read.csv("Common hESC-IMR90 boundaries.csv", header = FALSE)
ir.TAD = IRanges(start = TADbound[,2], end = TADbound[,3])
gr.bound = GRanges(ranges = ir.TAD, seqnames = TADbound[,1])

# gr.hESC.TAD is the set of hESC TADs alone as defined by Dixon (2012)
TADbound.hESC = read.csv("boundaries.hESC.csv", header = FALSE)
ir.TAD.hESC = IRanges(start = TADbound.hESC[,2], end = TADbound.hESC[,3])
gr.bound.hESC = GRanges(ranges = ir.TAD.hESC, seqnames = TADbound.hESC[,1])

# gr.hESC.TAD is the set of IMR90 TADs alone as defined by Dixon (2012)
TADbound.IMR90 = read.csv("boundaries.IMR90.csv", header = FALSE)
ir.TAD.IMR90 = IRanges(start = TADbound.IMR90[,2], end = TADbound.IMR90[,3])
gr.bound.IMR90 = GRanges(ranges = ir.TAD.IMR90, seqnames = TADbound.IMR90[,1])

# Where gr.enhs is all enhancer:promoter interactions 
load("/home/kaiti/Documents/R Data/gr.enhs.rda")
setwd("/home/kaiti/Documents/")
```

# TAD Boundary Crossings
To determine how many TAD boundaries are crossed by an enhnacer:promoter interaction
```{r}
chromlist = c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y')
```
```{r}
chromosome = c()
enhs.num = c()
shared.bound = c()
x.bound = c()
hESC.bound = c()
x.hESC = c()
hESC.only.bound = c()
x.hESC.only = c()
IMR90.bound = c()
x.IMR90 = c()
IMR90.only.bound = c()
x.IMR90.only = c()
boundFrame = data.frame(chromosome, enhs.num, shared.bound, x.bound, hESC.bound, x.hESC, hESC.only.bound, x.hESC.only, IMR90.bound, x.IMR90, IMR90.only.bound, x.IMR90.only)    
```
```{r}
UniqueBounds = function(i, frame, tads, enhs, hESC, IMR90){
  # Define TAD boundaries and enhancer:promoter interactions in a chromsome specific manner
  gr.bound = tads[seqnames(tads)==paste("chr",as.character(i),sep="")]
  ir.TAD = IRanges(start = start(gr.bound), end = end(gr.bound))
  
  h.bound = hESC[seqnames(hESC)==paste("chr",as.character(i),sep="")]
  h.bound = IRanges(start = start(h.bound), end = end(h.bound))
  I.bound = IMR90[seqnames(IMR90)==paste("chr",as.character(i),sep="")]
  I.bound = IRanges(start = start(I.bound), end = end(I.bound))
  gaps = gaps(ir.TAD)
  
  enhs.chr = enhs[seqnames(enhs)==paste("chr",as.character(i),sep="")]
  ir.enhs = IRanges(start = enhs.chr$chromStart, end = enhs.chr$chromEnd)

  # Defining Unique Boundaries
  IMR90.ov = findOverlapPairs(I.bound, gaps)
  max.ov.IMR = IRanges(start = pmin(start(first(IMR90.ov)), start(second(IMR90.ov))), end = pmax(end(first(IMR90.ov)), end(second(IMR90.ov))))
  IMR.bound.x = first(IMR90.ov)
  uniq.I.bound = unique(IMR.bound.x[start(IMR.bound.x) >= start(max.ov.IMR) & end(IMR.bound.x) <= end(max.ov.IMR)])
  ir.I.bound = IRanges(start = start(uniq.I.bound), end = end(uniq.I.bound))

  hESC.ov = findOverlapPairs(h.bound, gaps)
  max.ov.hESC = IRanges(start = pmin(start(first(hESC.ov)), start(second(hESC.ov))), end = pmax(end(first(hESC.ov)), end(second(hESC.ov))))
  hESC.bound.x = first(hESC.ov)
  uniq.h.bound = unique(hESC.bound.x[start(hESC.bound.x) >= start(max.ov.hESC) & end(hESC.bound.x) <= end(max.ov.hESC)])
  ir.h.bound = IRanges(start = start(uniq.h.bound), end = end(uniq.h.bound))
  
  # Collecting Interactions by Type 
  x = findOverlapPairs(ir.enhs, ir.TAD)
  x.tad = (second(findOverlapPairs(ir.enhs, ir.TAD)))
  max.ovlp = (IRanges(start = pmin(start(first(x)), start(second(x))), end = pmax(end(first(x)), end(second(x)))))
  x.TAD = unique(x.tad[start(x.tad) > start(max.ovlp) & end(x.tad) < end(max.ovlp)])
  
  # Interactions that cross cell-specific boundaries
  h = findOverlapPairs(ir.enhs, h.bound)
  hx = second(findOverlapPairs(ir.enhs, h.bound))
  h.max.ovlp = (IRanges(start = pmin(start(first(h)), start(second(h))), end = pmax(end(first(h)), end(second(h)))))
  hx.hESC = unique(hx[start(hx) > start(h.max.ovlp) & end(hx) < end(h.max.ovlp)])

  I = findOverlapPairs(ir.enhs, I.bound)
  Ix = second(findOverlapPairs(ir.enhs, I.bound))
  I.max.ovlp = (IRanges(start = pmin(start(first(I)), start(second(I))), end = pmax(end(first(I)), end(second(I)))))
  Ix.IMR90 = unique(Ix[start(Ix) > start(I.max.ovlp) & end(Ix) < end(I.max.ovlp)])
  
  h.only = findOverlapPairs(ir.enhs, ir.h.bound)
  hx.only = second(findOverlapPairs(ir.enhs, ir.h.bound))
  h.only.max.ovlp = (IRanges(start = pmin(start(first(h.only)), start(second(h.only))), end = pmax(end(first(h.only)), end(second(h.only)))))
  hx.hESC.only = unique(hx.only[start(hx.only) > start(h.only.max.ovlp) & end(hx.only) < end(h.only.max.ovlp)])

  I.only = findOverlapPairs(ir.enhs, ir.I.bound)
  Ix.only = second(findOverlapPairs(ir.enhs, ir.I.bound))
  I.only.max.ovlp = (IRanges(start = pmin(start(first(I.only)), start(second(I.only))), end = pmax(end(first(I.only)), end(second(I.only)))))
  Ix.IMR90.only = unique(Ix.only[start(Ix.only) > start(I.only.max.ovlp) & end(Ix.only) < end(I.only.max.ovlp)])
  
  # Storing all that new data
    chromosome = append(chromosome, as.character(i))
    enhs.num = append(enhs.num, length(ir.enhs))
    shared.bound = append(shared.bound, length(gr.bound))
    x.bound = append(x.bound, length(x.TAD))
    hESC.bound = append(hESC.bound, length(h.bound))
    x.hESC = append(x.hESC, length(hx.hESC))
    hESC.only.bound = append(hESC.only.bound, length(uniq.h.bound))
    x.hESC.only = append(x.hESC.only, length(hx.hESC.only))
    IMR90.bound = append(IMR90.bound, length(I.bound))
    x.IMR90 = append(x.IMR90, length(Ix.IMR90))
    IMR90.only.bound = append(IMR90.only.bound, length(uniq.I.bound))
    x.IMR90.only = append(x.IMR90.only, length(Ix.IMR90.only))
    
    frame = data.frame(chromosome, enhs.num, shared.bound, x.bound, hESC.bound, x.hESC, hESC.only.bound, x.hESC.only, IMR90.bound, x.IMR90, IMR90.only.bound, x.IMR90.only)          
    return(frame)
    }
```
```{r}
data = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=gr.enhs, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "TAD boundary Summary.csv")
```
^ Crossings in the above refer to TAD boundaries which are *fully* encompassed by at least one enhancer:promoter interaction.

# Summarizing Enhancer:Promoter Interactions Spanning, Encompassed by, or Touching TAD Boundaries. 
```{r}
chromosome = c()
TAD.num = c()
enhs.num = c()
enhs.interact = c()
enhs.encompassed = c()
enhs.engulfed = c()
chromFrame = data.frame(chromosome, TAD.num, enhs.num, enhs.interact, enhs.encompassed, enhs.engulfed)    
```
```{r}
ChromCross = function(i, frame, tads, enhs){
  # Define TAD boundaries and enhancer:promoter interactions in a chromsome specific manner
  gr.bound = tads[seqnames(tads)==paste("chr",as.character(i),sep="")]
  ir.TAD = IRanges(start = start(gr.bound), end = end(gr.bound))
  
  enhs.chr = enhs[seqnames(enhs)==paste("chr",as.character(i),sep="")]
  ir.enhs = IRanges(start = enhs.chr$chromStart, end = enhs.chr$chromEnd)

  # Collecting Interactions by Type 
  x = findOverlapPairs(ir.enhs, ir.TAD)
  x.enhs = first(findOverlapPairs(ir.enhs, ir.TAD))
  x.tad = (second(findOverlapPairs(ir.enhs, ir.TAD)))
  max.ovlp = (IRanges(start = pmin(start(first(x)), start(second(x))), end = pmax(end(first(x)), end(second(x)))))

  enhs.x = unique(x.enhs)
  TAD.x = unique(x.tad)
  enhs.full = unique(x.enhs[start(x.enhs)==start(max.ovlp)& end(x.enhs)==end(max.ovlp)])
  TAD.full = unique(x.tad[start(x.tad)==start(max.ovlp) & end(x.tad)==end(max.ovlp)])
  enhs.flanked = unique(x.enhs[start(x.enhs) > start(max.ovlp) & end(x.enhs) < end(max.ovlp)])
    
  # Storing all that new data
    chromosome = append(chromosome, as.character(i))
    TAD.num = append(TAD.num, length(ir.TAD))
    enhs.num = append(enhs.num, length(ir.enhs))
    enhs.interact = append(enhs.interact, length(enhs.x))
    enhs.encompassed = append(enhs.encompassed, length(enhs.full))
    enhs.engulfed = append(enhs.engulfed, length(enhs.flanked))
    
    frame = data.frame(chromosome, TAD.num, enhs.num, enhs.interact, enhs.encompassed, enhs.engulfed)  
    return(frame)
    }
```
```{r}
data = lapply(chromlist, ChromCross, frame=chromFrame, tads=gr.bound, enhs=gr.enhs)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "ENHS Interactions with Boundary Elements.csv")
```
^ Above file uses the entire gr.enhs data set and categorizes the manner in which the enhancer:promoter interactions cross a TAD boundary (either, an interaction (i.e. single sideded), encompassing the boundary element, or encompassed by the boundary element).


# Classifying Crossed Boundaries While Accounting for Boundary Size
```{r}
length(ir.TAD)
min(width(ir.TAD))
max(width(ir.TAD))
sum(width(ir.TAD)==80001)
sum(width(ir.TAD)==120001)
sum(width(ir.TAD)==160001)
sum(width(ir.TAD)==200001)
sum(width(ir.TAD)>200001)

length(ir.TAD.hESC)
min(width(ir.TAD.hESC))
max(width(ir.TAD.hESC))
sum(width(ir.TAD.hESC)==1)
sum(width(ir.TAD.hESC)==40001)
sum(width(ir.TAD.hESC)==80001)
sum(width(ir.TAD.hESC)==120001)
sum(width(ir.TAD.hESC)==160001)
sum(width(ir.TAD.hESC)==200001)
sum(width(ir.TAD.hESC)>200001)

length(ir.TAD.IMR90)
min(width(ir.TAD.IMR90))
max(width(ir.TAD.IMR90))
sum(width(ir.TAD.IMR90)==1)
sum(width(ir.TAD.IMR90)==40001)
sum(width(ir.TAD.IMR90)==80001)
sum(width(ir.TAD.IMR90)==120001)
sum(width(ir.TAD.IMR90)==160001)
sum(width(ir.TAD.IMR90)==200001)
sum(width(ir.TAD.IMR90)>200001)
```
based on the categorization of TAD boundary widths described above I've decided to initially proceed with the shared boundaries broken into bins of 80 kb, 120 kb, and >120 kb. For the cell-specific TAD boundaries I've decided to proceed by breaking into bins of width 1, 40 kb, 80 kb, 120 kb, and larger than 120 kb.
```{r}
gr.bound.80 = gr.bound[width(gr.bound)==80001]
gr.bound.120 = gr.bound[width(gr.bound)==120001]
gr.bound.lg = gr.bound[width(gr.bound)> 120001]

gr.bound.hESC.1 = gr.bound.hESC[width(gr.bound.hESC)==1]
gr.bound.hESC.40 = gr.bound.hESC[width(gr.bound.hESC)==40001]
gr.bound.hESC.80 = gr.bound.hESC[width(gr.bound.hESC)==80001]
gr.bound.hESC.120 = gr.bound.hESC[width(gr.bound.hESC)==120001]
gr.bound.hESC.lg = gr.bound.hESC[width(gr.bound.hESC)>120001]

gr.bound.IMR90.1 = gr.bound.IMR90[width(gr.bound.IMR90)==1]
gr.bound.IMR90.40 = gr.bound.IMR90[width(gr.bound.IMR90)==40001]
gr.bound.IMR90.80 = gr.bound.IMR90[width(gr.bound.IMR90)==80001]
gr.bound.IMR90.120 = gr.bound.IMR90[width(gr.bound.IMR90)==120001]
gr.bound.IMR90.lg = gr.bound.IMR90[width(gr.bound.IMR90)>120001]
```
```{r}
chromosome = c()
enhs.num = c()
bound.num = c()
bound.interact = c()
bound.crossed = c()
TADframe = data.frame(chromosome, enhs.num, bound.num, bound.interact, bound.crossed)
```
```{r}
XdBound = function(i, frame, enhs, TAD){
  # Define TAD boundaries and enhancer:promoter interactions in a chromsome specific manner
  chr.TAD = TAD[seqnames(TAD)==paste("chr",as.character(i),sep="")]
  ir.TAD = IRanges(start = start(chr.TAD), end = end(chr.TAD))
  
  enhs.chr = enhs[seqnames(enhs)==paste("chr",as.character(i),sep="")]
  ir.enhs = IRanges(start = enhs.chr$chromStart, end = enhs.chr$chromEnd)

  # Collecting Interactions by Type 
  x = findOverlapPairs(ir.enhs, ir.TAD)
  x.enhs = first(findOverlapPairs(ir.enhs, ir.TAD))
  x.tad = (second(findOverlapPairs(ir.enhs, ir.TAD)))
  max.ovlp = (IRanges(start = pmin(start(first(x)), start(second(x))), end = pmax(end(first(x)), end(second(x)))))

  TAD.interact = unique(x.tad)
  TAD.crossed = unique(x.tad[start(x.tad) > start(max.ovlp) & end(x.tad) < end(max.ovlp)])
    
  # Storing all that new data
    chromosome = append(chromosome, as.character(i))
    enhs.num = append(enhs.num, length(ir.enhs))
    bound.num = append(bound.num, length(ir.TAD))
    bound.interact = append(bound.interact, length(TAD.interact))
    bound.crossed = append(bound.crossed, length(TAD.crossed))
    
    frame = data.frame(chromosome, enhs.num, bound.num, bound.interact, bound.crossed)
    return(frame)
    }
```
```{r}
sum.function = function(x){
  DataSummary = data.frame()
  for(i in seq_along(x)){
    DataSummary = rbind(DataSummary, x[[i]])
  }
  return(DataSummary)
  }
```
```{r}
shared.all = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound)
  sum.shared.all = sum.function(shared.all)
  write.csv(sum.shared.all, file = "All Shared Boundaries.csv")
shared.80 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.80)
  sum.shared.80 = sum.function(shared.80)
  write.csv(sum.shared.80, file ="Shared Boundaries of 80kb.csv")
shared.120 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.120)
  sum.shared.120 = sum.function(shared.120)
  write.csv(sum.shared.120, file="Shared Boundaries of 120 kb.csv")
shared.lg = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.lg)
  sum.shared.lg = sum.function(shared.lg)
  write.csv(sum.shared.lg, file ="Shared Boundaries >120 kb.csv")

hESC.all = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.hESC)
  sum.hESC.all = sum.function(hESC.all)
  write.csv(sum.hESC.all, file = "All hESC Boundaries.csv")
hESC.1 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.hESC.1)
  sum.hESC.1 = sum.function(hESC.1)
  write.csv(sum.hESC.1, file = "hESC Boundaries of 1.csv")
hESC.40 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.hESC.40)
  sum.hESC.40 = sum.function(hESC.40)
  write.csv(sum.hESC.40, file = "Summary hESC Boundaries of 40 kb.csv")
hESC.80 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.hESC.80)
  sum.hESC.80 = sum.function(hESC.80)
  write.csv(sum.hESC.80, file = "Summary hESC Boundaries of 80 kb.csv")  
hESC.120 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.hESC.120)
  sum.hESC.120 = sum.function(hESC.120)
  write.csv(sum.hESC.120, file = "Summary hESC Boundaries of 120 kb.csv")
hESC.lg = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.hESC.lg)
  sum.hESC.lg = sum.function(hESC.lg)
  write.csv(sum.hESC.lg, file = "Summary hESC Boundaries > 120 kb.csv")

IMR90.all = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.IMR90)
  sum.IMR90.all = sum.function(IMR90.all)
  write.csv(sum.IMR90.all, file = "All IMR90 Boundaries.csv")
IMR90.1 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.IMR90.1)
  sum.IMR90.1 = sum.function(IMR90.1)
  write.csv(sum.IMR90.1, file = "IMR90 Boundaries of 1.csv")
IMR90.40 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.IMR90.40)
  sum.IMR90.40 = sum.function(IMR90.40)
  write.csv(sum.IMR90.40, file = "IMR90 Boundaries of 40 kb.csv")
IMR90.80 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.IMR90.80)
  sum.IMR90.80 = sum.function(IMR90.80)
  write.csv(sum.IMR90.80, file = "IMR90 Boundaries of 80 kb.csv")
IMR90.120 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.IMR90.120)
  sum.IMR90.120 = sum.function(IMR90.120)
  write.csv(sum.IMR90.120, file = "IMR90 Boundaries of 120 kb.csv")
IMR90.lg = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.IMR90.lg)
  sum.IMR90.lg = sum.function(IMR90.lg)
  write.csv(sum.IMR90.lg, file = "IMR90 Boundaries > 120 kb.csv")
```
```{r}
gr.bound.80 = gr.bound[width(gr.bound)==80001]
gr.bound.120 = gr.bound[width(gr.bound)==120001]
gr.bound.160 = gr.bound[width(gr.bound)==160001]
gr.bound.200 = gr.bound[width(gr.bound)==200001]
gr.bound.lg = gr.bound[width(gr.bound)> 200001]

shared.80 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.80)
  sum.shared.80 = sum.function(shared.80)

shared.120 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.120)
  sum.shared.120 = sum.function(shared.120)

shared.160 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.160)
  sum.shared.160 = sum.function(shared.160)
  
shared.200 = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.200)
  sum.shared.200 = sum.function(shared.200)
  
shared.lg = lapply(chromlist, XdBound, frame=TADframe, enhs=gr.enhs, TAD=gr.bound.lg)
  sum.shared.lg = sum.function(shared.lg)
  
sum.shared.all
sum.shared.80
sum.shared.120
sum.shared.160
sum.shared.200
```
Unsurprisingly, as the boundary elements increase in width, the % of boundaries which are crossed decreases. I'm going to set this analysis aside for now and will try to take a look at the number of gene ranges which cross a boundary element. 

# Analysis with RefSeq Gene Ranges
```{r}
data = lapply(chromlist, ChromCross, frame=chromFrame, tads=gr.bound, enhs=gr.hg19)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "Gene Interactions with Boundary Elements.csv")
```
Of the set of RefSeq genes, about 3000 (of the ~63000) are discarded from my analysis due to unclear chromsome names. Of the included genes approximately 10% interact with a TAD boundary, about 0.5% span a TAD boundary and arounf 6% are contained within a TAD boundary.

# Breaking gr.enhs down by score
The score for each range is a reflection of the maximum pooled expression of TCs used to construct the bidirectional loci. My inlinatinon is to expect that in this case, the genes that are more highly expressed, may be expressed in more tissues, and subsequently may be more likely to cross a TAD boundary. 
```{r}
gr.enhs.q1 = gr.enhs[score(gr.enhs)<221]
gr.enhs.q2 = gr.enhs[score(gr.enhs)>220 & score(gr.enhs)<300]
gr.enhs.q3 = gr.enhs[score(gr.enhs)>299 & score(gr.enhs)<440]
gr.enhs.q4 = gr.enhs[score(gr.enhs)>439]
```
How Many Enhancer Interactions in each quartile of scores cross a TAD boundary?
```{r}
data = lapply(chromlist, ChromCross, frame=chromFrame, tads=gr.bound, enhs=gr.enhs.q1)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "Score.Q1 ENHS Interactions with Boundary Elements.csv")
```
```{r}
data = lapply(chromlist, ChromCross, frame=chromFrame, tads=gr.bound, enhs=gr.enhs.q2)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "Score.Q2 ENHS Interactions with Boundary Elements.csv")
```
```{r}
data = lapply(chromlist, ChromCross, frame=chromFrame, tads=gr.bound, enhs=gr.enhs.q3)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "Score.Q3 ENHS Interactions with Boundary Elements.csv")
```
```{r}
data = lapply(chromlist, ChromCross, frame=chromFrame, tads=gr.bound, enhs=gr.enhs.q4)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
DataSummary
write.csv(DataSummary, file = "Score.Q4 ENHS Interactions with Boundary Elements.csv")
```
Conclusions: So, no, actually, as the score increases the number % of interactions that touch a TAD boundary and the % of interactions that span a TAD boundary increases, however, the % of interactions which are contained within a TAD boundary increases...which is unexpected
```{r}
par(mfrow= c(2,2))
hist(score(gr.enhs.q1))
hist(score(gr.enhs.q2))
hist(score(gr.enhs.q3))
hist(score(gr.enhs.q4))
```
```{r}
ir.enhs.q1 = IRanges(start = gr.enhs.q1$chromStart, end = gr.enhs.q1$chromEnd)
ir.enhs.q2 = IRanges(start = gr.enhs.q2$chromStart, end = gr.enhs.q2$chromEnd)
ir.enhs.q3 = IRanges(start = gr.enhs.q3$chromStart, end = gr.enhs.q3$chromEnd)
ir.enhs.q4 = IRanges(start = gr.enhs.q4$chromStart, end = gr.enhs.q4$chromEnd)

par(mfrow = c(2,2))
hist(width(ir.enhs.q1))
hist(width(ir.enhs.q2))
hist(width(ir.enhs.q3))
hist(width(ir.enhs.q4))
```
After looking at summaries of the range widths when broken down into quartiles I think the difference in percent of interactions that cross a boundary is due to the difference metrics of range width in the quartile bins. While all four quartiles cover approx. the same range of widths, the 1st and 2nd score-quartiles have a more even distribution across all widths. Longer interactions are of course going to be more likely to cross a TAD boundary. Since there is a smaller proportion of long ranges in the fourth score-quartile we shouldn't be surprised that a greater percentage ranges are encompassed by boundaries and a smaller percentage span boundaries. I think this is interesting in that the more highly scored interactions are less likely to be long-range interactions. I need to reivist the paper to better understand their methodology in defining an interaction interval. 

```{r}
data = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=gr.hg19, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)

DataSummary = data.frame()
for(i in seq_along(data)){
  DataSummary = rbind(DataSummary, data[[i]])
}
#DataSummary
geneData =DataSummary
write.csv(DataSummary, file = "TADs Crossed by Genes Summary.csv")
```

New Question: For the boundaries that are crossed by an enhancer:promoter interactions, how many times is the boundary crossed?

```{r}
NumCross = function(i, tads, enhs){
  # Define TAD boundaries and enhancer:promoter interactions in a chromsome specific manner
  gr.bound = tads[seqnames(tads)==paste("chr",as.character(i),sep="")]
  ir.TAD = IRanges(start = start(gr.bound), end = end(gr.bound))
  
  enhs.chr = enhs[seqnames(enhs)==paste("chr",as.character(i),sep="")]
  ir.enhs = IRanges(start = enhs.chr$chromStart, end = enhs.chr$chromEnd)

# Tallying the number of times a boundary is crossed
  x = findOverlapPairs(ir.enhs, ir.TAD)
  x.enhs = first(findOverlapPairs(ir.enhs, ir.TAD))
  x.tad = (second(findOverlapPairs(ir.enhs, ir.TAD)))
  max.ovlp = (IRanges(start = pmin(start(first(x)), start(second(x))), end = pmax(end(first(x)), end(second(x)))))

  TAD.interact = unique(x.tad)
  TAD.crossed = (x.tad[start(x.tad) > start(max.ovlp) & end(x.tad) < end(max.ovlp)])
  uniq.cross = unique(x.tad[start(x.tad) > start(max.ovlp) & end(x.tad) < end(max.ovlp)])

  counts = c()
  for(i in seq_along(uniq.cross)){
    counts = append(counts, sum(start(TAD.crossed)==start(uniq.cross)[i]))
  }
  return(counts)
}
```
```{r}
countdata = lapply(chromlist, NumCross,  tads=gr.bound, enhs=gr.enhs)
```

```{r}
one = NumCross(1,gr.bound,gr.enhs)
two = NumCross(2,gr.bound,gr.enhs)
three = NumCross(3,gr.bound,gr.enhs)
four = NumCross(4,gr.bound,gr.enhs)
five = NumCross(5,gr.bound,gr.enhs)
six = NumCross(6,gr.bound,gr.enhs)
seven = NumCross(7,gr.bound,gr.enhs)
eight = NumCross(8,gr.bound,gr.enhs)
nine = NumCross(9,gr.bound,gr.enhs)
ten = NumCross(10,gr.bound,gr.enhs)
eleven = NumCross(11,gr.bound,gr.enhs)
twelve = NumCross(12,gr.bound,gr.enhs)
thirteen = NumCross(13,gr.bound,gr.enhs)
fourteen = NumCross(14,gr.bound,gr.enhs)
fifteen = NumCross(15,gr.bound,gr.enhs)
sixteen = NumCross(16,gr.bound,gr.enhs)
seventeen = NumCross(17,gr.bound,gr.enhs)
eighteen = NumCross(18,gr.bound,gr.enhs)
nineteen = NumCross(19,gr.bound,gr.enhs)
twenty = NumCross(20,gr.bound,gr.enhs)
twentyone = NumCross(21,gr.bound,gr.enhs)
twentytwo = NumCross(22,gr.bound,gr.enhs)
eX = NumCross('X',gr.bound,gr.enhs)
counts = c(one,two,three,four,five,six,seven,eight,nine,ten,eleven,twelve,thirteen,fourteen,fifteen,sixteen,seventeen,eighteen,nineteen,twenty,twentyone,twentytwo,eX)
```
```{r}
summary(counts)
par(mfrow = c(1,1))
plot.new()
hist(counts,breaks = seq(0,max(counts),by = 1), main = 'All Boundaries',xlab = '# of Interactions Across Boundary', ylab = '# of Boundaries')
counts.1 = counts[counts!=1]
summary(counts.1)
hist(counts.1,breaks = seq(0,max(counts),by = 1), main = 'Boundaries Crossed More Than Once',xlab = '# of Interactions Across Boundary', ylab = '# of Boundaries')
counts.2 = counts[counts >2]
summary(counts.2)
hist(counts.2,breaks = seq(0,max(counts),by = 1), main = 'Boundaries Crossed More Than Twice',xlab = '# of Interactions Across Boundary', ylab = '# of Boundaries')
counts.4 = counts[counts >4]
summary(counts.4)
hist(counts.4,breaks = seq(0,max(counts),by = 1),main = 'Boundaries Crossed More Than Four Times',xlab = '# of Interactions Across Boundary', ylab = '# of Boundaries')
```

## TAD Boundary Crossings

```{r}
CrossedBound = function(i, tads, enhs){
  gr.bound = tads[seqnames(tads)==paste("chr",as.character(i),sep="")]
  ir.TAD = IRanges(start = start(gr.bound), end = end(gr.bound))
  
  enhs.chr = enhs[seqnames(enhs)==paste("chr",as.character(i),sep="")]
  ir.enhs = IRanges(start = enhs.chr$chromStart, end = enhs.chr$chromEnd)
  
  x = findOverlapPairs(ir.enhs, ir.TAD)
  x.enhs = first(findOverlapPairs(ir.enhs, ir.TAD))
  x.tad = (second(findOverlapPairs(ir.enhs, ir.TAD)))
  max.ovlp = (IRanges(start = pmin(start(first(x)), start(second(x))), end = pmax(end(first(x)), end(second(x)))))

  TAD.interact = unique(x.tad)
  TAD.crossed = (x.tad[start(x.tad) > start(max.ovlp) & end(x.tad) < end(max.ovlp)])
  uniq.cross = unique(x.tad[start(x.tad) > start(max.ovlp) & end(x.tad) < end(max.ovlp)])
  
  counts = c()
  for(j in seq_along(uniq.cross)){
    counts = append(counts, sum(start(TAD.crossed)==start(uniq.cross)[j]))
  }
  
  crossings = GRanges(ranges = uniq.cross, seqnames = paste("chr",as.character(i),sep=""), num.x = counts)
  return(crossings)
}

```
Crossed boundaries returns the number of boundaries that are crossed by a set of enhancer:pronoter interactions


Thought: If the 'R' in the mcol data on gr.enhs refers to the specifity score then enhancer interactions with a lower R value should be unspecified and will have a higher chance of crossing a TAD boundary than the specified boundaries
```{r}
quantile(na.omit(gr.enhs$R), probs = seq(0,1,0.2))
enhs.specQ1 = gr.enhs[na.omit(gr.enhs$R <= 0.208)]
enhs.specQ2 = gr.enhs[na.omit(gr.enhs$R > 0.208 & gr.enhs$R <= 0.262)]
enhs.specQ3 = gr.enhs[na.omit(gr.enhs$R > 0.262 & gr.enhs$R <= 0.344)]
enhs.specQ4 = gr.enhs[na.omit(gr.enhs$R > 0.344 & gr.enhs$R <= 0.485)]
enhs.specQ5 = gr.enhs[na.omit(gr.enhs$R > 0.485)]
```
```{r}
Rscore.Q1 = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=enhs.specQ1, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)
  sum.Rscore.Q1 = summary(Rscore.Q1)
  write.csv(sum.Rscore.Q1, file = "Boundaried crossed by R-score Q1 enhancer interactions.csv")
Rscore.Q2 = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=enhs.specQ2, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)
  sum.Rscore.Q2 = summary(Rscore.Q2)
  write.csv(sum.Rscore.Q2, file = "Boundaried crossed by R-score Q2 enhancer interactions.csv")
Rscore.Q3 = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=enhs.specQ3, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)
  sum.Rscore.Q3 = summary(Rscore.Q3)
  write.csv(sum.Rscore.Q3, file = "Boundaried crossed by R-score Q3 enhancer interactions.csv")
Rscore.Q4 = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=enhs.specQ4, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)
  sum.Rscore.Q4 = summary(Rscore.Q4)
  write.csv(sum.Rscore.Q4, file = "Boundaried crossed by R-score Q4 enhancer interactions.csv")
Rscore.Q5 = lapply(chromlist, UniqueBounds, frame=boundFrame, tads=gr.bound, enhs=enhs.specQ5, hESC=gr.bound.hESC, IMR90=gr.bound.IMR90)
  sum.Rscore.Q5 = summary(Rscore.Q5)
  write.csv(sum.Rscore.Q5, file = "Boundaried crossed by R-score Q5 enhancer interactions.csv")  
```

```{r}
enhs21 = gr.enhs[seqnames(gr.enhs)=="chr21"]
bound21 = gr.bound[seqnames(gr.bound)=="chr21"]
ir.enhs21 = IRanges(start = enhs21$chromStart, end = enhs21$chromEnd)
ir.bound21 = IRanges(start = start(bound21), end = end(bound21))
a = unique(second(findOverlapPairs(ir.enhs21,ir.bound21)))
b = unique(first(findOverlapPairs(ir.enhs21,ir.bound21)))

```